src/test/java/org/jenkinsci/plugins/workflow/graphanalysis/FlowScannerTest.java

```diff
@@ -0,0 +1,492 @@
+         2 - FlowStartNode (BlockStartNode)
+         3 - Echostep
+         4 - ExecutorStep (StepStartNode) - WorkspaceAction
+         5 - ExecutorStep (StepStartNode) - BodyInvocationAction
+         6 - Echostep
+         7 - StepEndNode - startId (5)
+         8 - StepEndNode - startId (4)
+         9 - FlowEndNode
+         */
+
+        WorkflowRun b = r.assertBuildStatusSuccess(job.scheduleBuild2(0));
+        Collection<FlowNode> heads = b.getExecution().getCurrentHeads();
+        FlowExecution exec = b.getExecution();
+
+        LinearBlockHoppingScanner hopper = new LinearBlockHoppingScanner();
+        FlowNode headCandidate = exec.getNode("7");
+        Assert.assertEquals(exec.getNode("4"), hopper.jumpBlockScan(headCandidate, Collections.EMPTY_SET));
+        Assert.assertTrue("Setup should return true if we can iterate", hopper.setup(headCandidate, null));
+
+        headCandidate = exec.getNode("6");
+        List<FlowNode> filtered = hopper.filteredNodes(headCandidate, MATCH_ECHO_STEP);
+        Assert.assertEquals(2, filtered.size());
+
+        headCandidate = exec.getNode("7");
+        filtered = hopper.filteredNodes(Collections.singleton(headCandidate), null, MATCH_ECHO_STEP);
+        Assert.assertEquals(1, filtered.size());
+
+        filtered = hopper.filteredNodes(Collections.singleton(exec.getNode("8")), null, MATCH_ECHO_STEP);
+        Assert.assertEquals(1, filtered.size());
+
+        filtered = hopper.filteredNodes(Collections.singleton(exec.getNode("9")), null, MATCH_ECHO_STEP);
+        Assert.assertEquals(0, filtered.size());
+    }
+
+
+    /** And the parallel case */
+    @Test
+    public void testParallelScan() throws Exception {
+        WorkflowJob job = r.jenkins.createProject(WorkflowJob.class, "Convoluted");
+        job.setDefinition(new CpsFlowDefinition(
+            "echo 'first'\n" +
+            "def steps = [:]\n" +
+            "steps['1'] = {\n" +
+            "    echo 'do 1 stuff'\n" +
+            "}\n" +
+            "steps['2'] = {\n" +
+            "    echo '2a'\n" +
+            "    echo '2b'\n" +
+            "}\n" +
+            "parallel steps\n" +
+            "echo 'final'"
+        ));
+
+        /** Flow structure (ID - type)
+         2 - FlowStartNode (BlockStartNode)
+         3 - Echostep
+         4 - ParallelStep (StepStartNode) (start branches)
+         6 - ParallelStep (StepStartNode) (start branch 1), ParallelLabelAction with branchname=1
+         7 - ParallelStep (StepStartNode) (start branch 2), ParallelLabelAction with branchname=2
+         8 - EchoStep, (branch 1) parent=6
+         9 - StepEndNode, (end branch 1) startId=6, parentId=8
+         10 - EchoStep, (branch 2) parentId=7
+         11 - EchoStep, (branch 2) parentId = 10
+         12 - StepEndNode (end branch 2)  startId=7  parentId=11,
+         13 - StepEndNode (close branches), parentIds = 9,12, startId=4
+         14 - EchoStep
+         15 - FlowEndNode (BlockEndNode)
+         */
```

Comment:
I think somewhere I mentioned that for the future it would be useful to create test fixtures which allow you to assert something like the above comment directly.